### ISSUE TITLE
Regression on 1.11 - Modular Implementation + Definition

### ISSUE THREAD

**Issue by GedMarc** [raphw/byte-buddy#1081]
Hey Guys,

Please confirm - 
* Added direct class dependency to findbugs on class net.bytebuddy.utility.dispatcher.JavaDispatcher
* Module file does not specify requirements for findbugs, 
* Resulting in class load error in modular execution mode

Violating line - 
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

Cause - 
No requires specified in module-info for findbugs, class is inaccessible 

JDK 16
Win 10
ByteBuddy 1.11 range 

Stacktrace 
[2021/07/13 00:39:31]-[Unable to get entity factory : net/bytebuddy/utility/dispatcher/JavaDispatcherjava.lang.NoClassDefFoundError: net/bytebuddy/utility/dispatcher/JavaDispatcher
	at net.bytebuddy/net.bytebuddy.description.type.TypeDescription$ForLoadedType.<clinit>(TypeDescription.java:8378)
	at net.bytebuddy/net.bytebuddy.description.type.TypeDescription.<clinit>(TypeDescription.java:65)
	at net.bytebuddy/net.bytebuddy.matcher.ElementMatchers.isFinalizer(ElementMatchers.java:1626)
	at net.bytebuddy/net.bytebuddy.matcher.ElementMatchers.isDefaultFinalizer(ElementMatchers.java:1616)
	at net.bytebuddy/net.bytebuddy.ByteBuddy.<init>(ByteBuddy.java:207)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.bytecode.internal.bytebuddy.ByteBuddyState.<init>(ByteBuddyState.java:79)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.bytecode.internal.bytebuddy.BytecodeProviderImpl.<init>(BytecodeProviderImpl.java:67)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.bytecode.internal.bytebuddy.BytecodeProviderImpl.<init>(BytecodeProviderImpl.java:58)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.cfg.Environment.buildBytecodeProvider(Environment.java:345)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.cfg.Environment.buildBytecodeProvider(Environment.java:337)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.cfg.Environment.<clinit>(Environment.java:230)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl$MergedSettings.<init>(EntityManagerFactoryBuilderImpl.java:1514)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl$MergedSettings.<init>(EntityManagerFactoryBuilderImpl.java:1504)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.mergeSettings(EntityManagerFactoryBuilderImpl.java:496)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:229)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:187)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:75)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilder(HibernatePersistenceProvider.java:171)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:119)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:61)
	at org.hibernate.orm.core@0.0.0_0-SNAPSHOT/org.hibernate.jpa.HibernatePersistenceProvider.createEntityManagerFactory(HibernatePersistenceProvider.java:50)
	at jakarta.persistence@0.0.0_0-SNAPSHOT/jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at com.guicedee.guicedpersistence@0.0.0_0-SNAPSHOT/com.guicedee.guicedpersistence.injectors.CustomJpaPersistService.start(CustomJpaPersistService.java:160)
	at com.guicedee.guicedpersistence@0.0.0_0-SNAPSHOT/com.guicedee.guicedpersistence.implementations.EntityManagerPostStartup.lambda$postLoad$0(EntityManagerPostStartup.java:43)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec$$$capture(ForkJoinTask.java:295)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassNotFoundException: net.bytebuddy.utility.dispatcher.JavaDispatcher
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:636)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:182)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
	... 35 more

**Comment by raphw** [comment]
I think the Hibernate team is already aware of this. It's not a bug in Byte Buddy, but likely an issue in how modules are used by Hibernate's class loaders. As for the issues:

1. *findbugs* is not included as it is not required at runtime. It's merely included to allow marking false-positive issues in Byte Buddy that should not be flagged during compilation. After compilation, the annotations do no longer serve a purpose.
2. `JavaDispatcher` should not be exposed as it is considered internal API.

**Comment by GedMarc** [comment]
Hey thanks - 

Much appreciate the reply :) 

Please check on line 18 - of the JavaDispatcher class - 
Unfortunately currently the class definitely has a reference to the findbugs @SuppressFBWarnings class, causing modular creation to fail, 

I have created a modular shade in the meantime so I can continue, there was also two changes in the module-info file, 

Confirmed this as temporary the fix for me, as to not alter the class itself in a shade or break anything
* Shade in findbugs (to avoid duplicate package detection in modules move package as well), 
* Update the module-info accordingly (exposing the package for JavaDispatcher as it is not currently exposed either, and remove any statics)

You can see my module-info overwrite and the shade specifications here 

https://github.com/GedMarc/GuicedEE-Services/tree/master/bytebuddy

![image](https://user-images.githubusercontent.com/5367513/125478598-09cd005e-2503-4ee2-ac62-f1ba89be4f64.png)

**Comment by GedMarc** [comment]
Looking forward to a legitimate fix so I can kill the shade :)

**Comment by raphw** [comment]
Not quite sure what you suggest.

The JVM allows for annotation types to be missing at runtime, where the annotation types are simply suppressed. This has been in the JVM since Java 5.

Why would it be needed to add a static dependency to findbugs here?

**Comment by GedMarc** [comment]
Oh sorry ;)

Suggest doing the following - 
* Removing the line in the JavaDispatcher class file to findbugs @FBSuppressWarnings to make sure its not a runtime dependency (it is definitely there in the class making it required, and breaking boot sequence in modular mode)
* Change the JNA requires in the module-info file to static to mark it as an optional runtime dependency 
* Expose the package for JavaDispatcher in the module-info file to allow access to it from other libraries

This looks to resolve everything in a modular environment

**Comment by GedMarc** [comment]
The JVM allowing an import for a class that is not found in the imports list is definitely not happening at all, the class fails on load due to a missing class,
When adding the class into the chain, It complains that it is not in the required set, 
When adding findbugs to the required set, it complains that the package is not exposed

When JNA is not on the module-path it complains that it is a required module as specified,

Either way I have a workaround and a working build for everything now, Up to you to bring it in, or believe me I guess xD xD

**Comment by raphw** [comment]
I found the issue, the package for *net.bytebuddy.utility.dispatcher* was missing from the module-info, where it needs to be mentioned, also without an export, to claim ownership.

JNA is already mentioned as a static requirement, how do you think it is required today? Here's what *javap* mentiones:

```java
module net.bytebuddy@1.11.7-SNAPSHOT {
  requires static java.instrument;
  requires static jdk.unsupported;
  requires static net.bytebuddy.agent;
  requires static com.sun.jna;
  requires static com.sun.jna.platform;
  // ...
}
```

Findbugs is only included as an annotations package with *compile* scope. It should not interfere with the module graph as the dependency is not available at runtime. What error do you find with it?

**Comment by raphw** [comment]
This should be fixed with version 1.11.7. Could you try the new release? As for findbugs, it does not even offer a stable module name, therefore including it is not really an option. But as said, only runtime annotations should be processed. If you get compile warnings about missing annotations, you need to include findbugs as a compile-time dependency to get rid of this.
